omegaconf/base.py

```diff
@@ -325,117 +311,194 @@ def _select_impl(
             throw_on_type_error=throw_on_resolution_failure,
         )
         if value is None:
-            return root, last_key, value
+            return root, last_key, None
         value = root._resolve_interpolation(
+            parent=root,
             key=last_key,
             value=value,
             throw_on_missing=False,
             throw_on_resolution_failure=True,
         )
         return root, last_key, value
 
-    def _resolve_simple_interpolation(
+    def _resolve_complex_interpolation(
         self,
+        parent: Optional["Container"],
+        value: "Node",
         key: Any,
-        inter_type: str,
+        parse_tree: OmegaConfGrammarParser.ConfigValueContext,
+        throw_on_missing: bool,
+        throw_on_resolution_failure: bool,
+    ) -> Optional["Node"]:
+        """
+        A "complex" interpolation is any interpolation that cannot be handled by
+        either `_resolve_node_interpolation()` or `_evaluate_custom_resolver()`, i.e.,
+        that either contains nested interpolations or is not a single "${..}" block.
+        """
+
+        from .nodes import StringNode
+
+        resolved = self.resolve_parse_tree(
+            parse_tree=parse_tree,
+            key=key,
+            parent=parent,
+            throw_on_missing=throw_on_missing,
+            throw_on_resolution_failure=throw_on_resolution_failure,
+        )
+
+        if resolved is None:
+            return None
+        elif isinstance(resolved, str):
+            # Result is a string: create a new node to store it.
```

Comment:
1. Changed as suggested in 8e7e205
2. This essentially corresponds to https://github.com/omry/omegaconf/blob/96645efc0af7fa0358abf4502ee744d06325e05d/omegaconf/base.py#L214 in the "old" way of doing interpolations. This is for string interpolations, e.g. `"foo_${bar}"`. That being said, thinking about it, it may be cleaner to create this `StringNode` within the grammar visitor itself (though I remember you preferred to minimize the amount of node-specific code in the visitor, so I'll only do it if you also think it'd be better)

Reply:
> That being said, thinking about it, it may be cleaner to create this `StringNode` within the grammar visitor itself (though I remember you preferred to minimize the amount of node-specific code in the visitor, so I'll only do it if you also think it'd be better)

Just a note that actually, in my current WIP work to address #488, I went more the opposite direction... I would suggest to resolve this point and discuss it again when I submit this follow-up work.

Reply:
sounds good.